src/NzbDrone.Mono/Disk/DiskProvider.cs

```diff
@@ -72,13 +72,49 @@ public override void InheritFolderPermissions(string filename)
             }
         }
 
-        public override void SetPermissions(string path, string mask, string user, string group)
+        public override void SetPermissions(string path, string mask)
         {
-            SetPermissions(path, mask);
-            SetOwner(path, user, group);
+            Logger.Debug("Setting permissions: {0} on {1}", mask, path);
+
+            var permissions = NativeConvert.FromOctalPermissionString(mask);
+
+            if (Directory.Exists(path))
+            {
+                permissions = GetFolderPermissions(permissions);
+            }
+
+            if (Syscall.chmod(path, permissions) < 0)
+            {
+                var error = Stdlib.GetLastError();
+
+                throw new LinuxPermissionsException("Error setting permissions: " + error);
+            }
+        }
+
+        private static FilePermissions GetFolderPermissions(FilePermissions permissions)
+        {
+            permissions |= (FilePermissions) ((int) (permissions & (FilePermissions.S_IRUSR | FilePermissions.S_IRGRP | FilePermissions.S_IROTH)) >> 2);
+
+            return permissions;
+        }
+
+        public override bool IsValidFilePermissionMask(string mask)
+        {
+            try
+            {
+                var permissions = NativeConvert.FromOctalPermissionString(mask);
+                return (permissions & (FilePermissions.S_ISGID | FilePermissions.S_ISUID | FilePermissions.S_ISVTX |
```

Comment:
I had to read it twice to understand what you were really checking, so probably better to split this up.
Separately check for setuid/setuid/sticky, return false otherwise.
Separately check no execute flags, return false otherwise.
Then check for rw for user.
That should make it easier to follow.